extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,622 @@
+    this.isFluid_ = isFluid;
+
+    /** @private {?({width: number, height: number}|../../../src/layout-rect.LayoutRectDef)} */
+    this.initialSize_ = initialSize;
+
+    /** @private {?({width, height}|../../../src/layout-rect.LayoutRectDef)} */
+    this.creativeSize_ = creativeSize;
+
+    /** @private {?string} */
+    this.fluidImpressionUrl_ = fluidImpressionUrl;
+
+    /** @private {boolean} */
+    this.sendPositionUpdate_ = false;
+
+    /** @private {?Promise} */
+    this.delay_ = null;
+
+    /** @private {../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = this.baseInstance_.getViewport();
+
+    /** @private {boolean} */
+    this.isCollapsed_ = true;
+
+    /** @private {boolean} */
+    this.isRegistered_ = false;
+
+    const sfConfig = Object(tryParseJson(
+        this.baseInstance_.element.getAttribute(
+            'data-safeframe-config')) || {});
+    /** @private {boolean} */
+    this.expandByOverlay_ = (sfConfig.hasOwnProperty('expandByOverlay') &&
+                             !!sfConfig['expandByOverlay']) || true;
+
+    /** @private {boolean} */
+    this.expandByPush_ = (sfConfig.hasOwnProperty('expandByPush') &&
+                          !!sfConfig['expandByPush']) || true;
+
+    this.registerSafeframeHost();
+  }
+
+  /**
+   * Returns the Safeframe specific name attributes that are needed for the
+   * Safeframe creative to properly setup.
+   * @return {!JsonObject}
+   */
+  getSafeframeNameAttr() {
+    const attributes = dict({});
+    attributes['uid'] = this.uid_;
+    attributes['hostPeerName'] = this.win_.location.origin;
+    attributes['initialGeometry'] = this.getInitialGeometry();
+    attributes['permissions'] = JSON.stringify(
+        dict({
+          'expandByOverlay': this.expandByOverlay_,
+          'expandByPush': this.expandByPush_,
+          'readCookie': false,
+          'writeCookie': false,
+        }));
+    attributes['metadata'] = JSON.stringify(
+        dict({
+          'shared': {
+            'sf_ver': this.baseInstance_.safeframeVersion,
+            'ck_on': 1,
+            'flash_ver': '26.0.0',
+          },
+        }));
+    attributes['reportCreativeGeometry'] = this.isFluid_;
+    attributes['isDifferentSourceWindow'] = false;
+    attributes['sentinel'] = this.sentinel_;
+    return attributes;
+  }
+
+  /**
+   * Returns the initialGeometry to assign to the name of the safeframe
+   * for rendering. This needs to be done differently than all the other
+   * geometry updates, because we don't actually have access to the
+   * rendered safeframe yet. Note that we are using getPageLayoutBox,
+   * which is not guaranteed to be perfectly accurate as it is from
+   * the last measure of the element. This is fine for our use case
+   * here, as even if the position is slightly off, we'll send the right
+   * size.
+   * @return {string}
+   */
+  getInitialGeometry() {
+    const ampAdBox = this.baseInstance_.element.getPageLayoutBox();
```

Comment:
nit: `this.baseInstance_.getPageLayoutBox()` will do

Reply:
done